File: Lainnya/movie-catalogue-jetpack-2-main/movie-catalogue-jetpack-2-main/movie-catalogue-jetpack-2-main/app/src/main/java/com/mramirid/moviecatalogue/viewmodel/ViewModelFactory.java
package com.mramirid.moviecatalogue.viewmodel;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.mramirid.moviecatalogue.data.source.MovieCatalogueRepository;
import com.mramirid.moviecatalogue.di.Injection;
import com.mramirid.moviecatalogue.ui.detail.DetailItemViewModel;
import com.mramirid.moviecatalogue.ui.movies.MoviesViewModel;
import com.mramirid.moviecatalogue.ui.tvshows.TvShowsViewModel;

public class ViewModelFactory extends ViewModelProvider.NewInstanceFactory {

	private static volatile ViewModelFactory INSTANCE;

	private final MovieCatalogueRepository movieCatalogueRepository;

	private ViewModelFactory(MovieCatalogueRepository movieCatalogueRepository) {
		this.movieCatalogueRepository = movieCatalogueRepository;
	}

	public static ViewModelFactory getInstance() {
		if (INSTANCE == null) {
			synchronized (ViewModelFactory.class) {
				if (INSTANCE == null)
					INSTANCE = new ViewModelFactory(Injection.provideRepository());
			}
		}
		return INSTANCE;
	}

	@NonNull
	@Override
	public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
		if (modelClass.isAssignableFrom(DetailItemViewModel.class)) {
			//noinspection unchecked
			return (T) new DetailItemViewModel(movieCatalogueRepository);
		} else if (modelClass.isAssignableFrom(MoviesViewModel.class)) {
			//noinspection unchecked
			return (T) new MoviesViewModel(movieCatalogueRepository);
		} else if (modelClass.isAssignableFrom(TvShowsViewModel.class)) {
			//noinspection unchecked
			return (T) new TvShowsViewModel(movieCatalogueRepository);
		}

		throw new IllegalArgumentException("Unknown ViewModel class named " + modelClass.getName());
	}
}

File: Lainnya/cielo/JetpackSubmission3v2/app/src/main/java/com/gracielo/jetpacksubmission3v2/ViewModelFactory.kt
package com.gracielo.jetpacksubmission3v2

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.gracielo.jetpacksubmission3v2.Data.Source.MovieCatalogueRepository
import com.gracielo.jetpacksubmission3v2.DataInjection.Injection
import com.gracielo.jetpacksubmission3v2.Detail.DetailViewModel
import com.gracielo.jetpacksubmission3v2.Favourite.MoviesFavouriteViewModel
import com.gracielo.jetpacksubmission3v2.Favourite.TVFavourtieViewModel
import com.gracielo.jetpacksubmission3v2.Movies.MoviesViewModel
import com.gracielo.jetpacksubmission3v2.TV.TVViewModel

class ViewModelFactory private constructor(private val movieCatalogueRepository: MovieCatalogueRepository) :
    ViewModelProvider.NewInstanceFactory() {
    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(DetailViewModel::class.java)) {
            return DetailViewModel(movieCatalogueRepository) as T
        } else if (modelClass.isAssignableFrom(MoviesViewModel::class.java)) {
            return MoviesViewModel(movieCatalogueRepository) as T
        } else if (modelClass.isAssignableFrom(TVViewModel::class.java)) {
            return TVViewModel(movieCatalogueRepository) as T
        } else if (modelClass.isAssignableFrom(MoviesFavouriteViewModel::class.java)) {
            return MoviesFavouriteViewModel(movieCatalogueRepository!!) as T
        }
        else if (modelClass.isAssignableFrom(TVFavourtieViewModel::class.java)) {
            return TVFavourtieViewModel(movieCatalogueRepository!!) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class named " + modelClass.name)
    }

    companion object {
        @Volatile
        private var INSTANCE: ViewModelFactory? = null
        fun getInstance(context: Context): ViewModelFactory =
            INSTANCE ?: synchronized(this) {
                ViewModelFactory(Injection.provideRepository(context)).apply { INSTANCE = this }
            }
    }
}